src/doom/r_data.c

```diff
@@ -1196,6 +1196,26 @@ void R_InitColormaps (void)
 	byte *const playpal = W_CacheLumpName("PLAYPAL", PU_STATIC);
 	byte *const colormap = W_CacheLumpName("COLORMAP", PU_STATIC);
 
+	// [crispy] Smoother diminishing lighting.
+	// Compiled in but not enabled TrueColor mode
+	// can't use more than original 32 colormaps.
+	if (crispy->truecolor && crispy->smoothlight)
+	{
+		NUMCOLORMAPS = 256;
+	}
+	else
+	{
+		NUMCOLORMAPS = 32;
+	}
+
+	// [crispy] zero out colormaps[] array so it can be
+	// reallocated and recalculated with various amount of colormaps
+	if (colormaps != NULL)
```

Comment:
TODO, CRITICAL: this doesn't seems to work correctly by preventing gamma-correction to work while in smoothest lighting. Possible solution:
* Use `malloc`/`free` to free `colormaps[]` array immediately. Doesn't seems to be safe by itself, as toggling smooth lighting on/off and changing gamma-correction may eventually lead to program crash, unless always allocate `256 + 1` instead of `NUMCOLORMAPS + 1` in `Z_Malloc` bellow.

Patch, friendly with memory consumption with smooth lighting on / off:
```diff
 	// [crispy] zero out colormaps[] array so it can be
 	// reallocated and recalculated with various amount of colormaps
-	if (colormaps != NULL)
+	if (colormaps)
 	{
 		free(colormaps);
-		colormaps = NULL;
 	}
 
-	if (!colormaps)
-	{
-		colormaps = (lighttable_t*) malloc((NUMCOLORMAPS + 1) * 256 * sizeof(lighttable_t));
-	}
+	colormaps = (lighttable_t*) malloc((256 + 1) * 256 * sizeof(lighttable_t));
```

Reply:
I don't quite get that code change. Why should we always allocate memory for 256 colormaps even if we explicitly have them set to 32? And let's please just use `I_Realloc()` here instead of that `free()`/`malloc()` dance. Thank you!

Reply:
I apologize. I have overheated a bit while code grinding, of course it should be just:
```
colormaps = I_Realloc (colormaps, (NUMCOLORMAPS + 1) * 256 * sizeof(lighttable_t));
```

Reply:
👍

Reply:
Nice to know in this context:

> In the GNU C Library, if the new size is the same as the old, realloc and reallocarray are guaranteed to change nothing and return the same address that you gave. However, POSIX and ISO C allow the functions to relocate the object or fail in this situation.

https://www.gnu.org/software/libc/manual/html_node/Changing-Block-Size.html

So, at least one implementation does the right thing by default. 😁

Reply:
I was trying to call `R_InitLightTables()` and `R_ExecuteSetViewSize()` for arrays recalculation while gamma changing, and it seems to fixing issue. But this was obliviously silly, turning simple gamma code into something too overcomplicated. 😶 Oh well.

Have a bad feeling about Hexen, it seems to have nearly same issue I have bumped while initial TrueColor support - brightest colormaps gets broken, but I'll investigate, no need to panic yet.

Reply:
> Have a bad feeling about Hexen, it seems to have nearly same issue I have bumped while initial TrueColor support - brightest colormaps gets broken, but I'll investigate, no need to panic yet.

Solved by using `I_Realloc`, so it was a right thing to do!